### ISSUE TITLE
Singularity build 

### ISSUE THREAD

**Issue by suujia** [bcgsc/orca#57]
[shub:bcgsc/orca](https://www.singularity-hub.org/containers/1167/log)
[shub:suujia/orca](https://www.singularity-hub.org/containers/3402/log)

Not sure how to trigger a rebuild for the bcgsc/orca shub. The error message I got with a build on suujia/orca shub was: 
```
+ su -c brew install perl -v linuxbrew
Error: /home/linuxbrew/.linuxbrew/Homebrew is not writable. You should change the
ownership and permissions of /home/linuxbrew/.linuxbrew/Homebrew back to your
user account:
sudo chown -R $(whoami) /home/linuxbrew/.linuxbrew/Homebrew
Error: Cannot write to /home/linuxbrew/.linuxbrew
ABORT: Aborting with RETVAL=255
```

I think I'll start with getting it to build locally first and updating singularity file to align with orca dockerfile.

**Comment by suujia** [comment]
I was able to use overlay to install nano and it's only available within this overlay. 
```
sudo singularity shell --overlay my-overlay.img orca.simg
```
However, when installing anything with brew inside:

```
Error: Running Homebrew as root is extremely dangerous and no longer supported.
As Homebrew does not drop privileges on installation you would be giving all
build scripts full access to your system.
```

**Comment by sjackman** [comment]
Excellent! That's good progress. The goal with using Singularity is that it shouldn't require root access for the user to run the image. It does I believe require root access to build the original Singularity image, but hopefully not for the user. Can you it without `sudo`?

**Comment by suujia** [comment]
```
singularity shell --overlay my-overlay.img orca.simg
```
Yes, after running this command, in directory where I don't have sudo access on the outside, I was able to `brew install __ ` things into the overlay image. This was with a small orca.simg I built. I am still debugging to build the whole orca singularity image. 
After it's built, I can start writing a script that creates an account with the individual user, most importantly: `singularity image.create --size 3000 $USER_overlay.simg`,    (note that our Singularity file is prebuilt into orca.simg, which can be run with no setup time) 

Users would just have to: 
```
ssh orca01
singularity shell --overlay $USER_overlay.simg orca.simg
```

Users not affiliated with bcgsc can just clone singularity locally, and then: 
```
singularity pull shub://bcgsc/orca
```

**Comment by sjackman** [comment]
Excellent progress! Thanks, Susie.

> `singularity image.create --size 3000 $USER_overlay.simg`

I'd suggest storing the overlay at `$HOME/.orca/overlay.simg`

**Comment by suujia** [comment]
Brew can't be run as sudo (due to error above) which is why it's ran as linuxbrew user: 
``` 
su -c 'brew update' linuxbrew
```

I have added `chown -R linuxbrew: /home/linuxbrew `  (I can change it later to add .linuxbrew), and it is rebuilding to same link. It is still building currently. I was running into the problem of individual sub directories requiring a chmod earlier. If it doesn't pass I will try with a chmod.

**Comment by sjackman** [comment]
> Brew can't be run as sudo (due to error above) which is why it's ran as linuxbrew user:

Ah, got it. Can the whole script be run as the Linuxbrew user rather than `root`? Or separate it into two parts, the `root` part and then then `linuxbrew` part, and switch users between those two parts. See here for an example in a Dockerfile. I'm not sure how this works in Singularity.

https://github.com/Linuxbrew/brew/blob/master/Dockerfile#L18

**Comment by suujia** [comment]
Yes, do linuxbrew user things in `su -c '__' linuxbrew` and just run the root commands outside with no wrappers. We are already doing that in this our Singularity file, with the brew commands run as linuxbrew user and everything else without it. Which other commands did you want to run under sudo? 

This is a really good example: [UofU bioBakery](https://github.com/CHPC-UofU/Singularity-bioBakery/blob/731317ca537962913789d817e4270f0027de03c7/Singularity#L79)

**Comment by sjackman** [comment]
Excellent find with the bioBakery Singularity file! That's a very useful resource.

`su -c '…' linuxbrew` is fine. I misunderstood what was going on.

Putting the option at the end may be more readable.
```
su linuxbrew -c '…'
```

**Comment by suujia** [comment]
### Bind Mounts:
Map directories on host system to directories within container (read/write data on host system). The following should be added to config file of singularity:  
```
ENABLE OVERLAY (boolean, default=’no’)  --  set to yes 
USER BIND CONTROL (boolean, default=’yes’) -- already yes 
```
====================
`pip3 install ruamel.yaml` fixes: 
```
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-install-kobo8rt4/ruamel.yaml/
```
 `cwlref-runner \` [here](https://pypi.org/project/cwltool/)
```
cwltool 1.0.20180711112827 has requirement ruamel.yaml<0.15,>=0.12.4, but you'll have ruamel-yaml 0.15.44 which is incompatible.
```
and 

```
    pysam/libchtslib.c:15077:12: error: too many arguments to function ‘(PyObject * (*)(PyObject *, PyObject * const*, Py_ssize_t))meth’
         return (*((__Pyx_PyCFunctionFast)meth)) (self, args, nargs, NULL);
```

**Comment by suujia** [comment]
`$PREFIX/etc/ singularity/singularity.conf.`(owned by root) should have following configurations set up. These configurations should be set up to control all the users using singularity command. However, these configs do not apply for root. 

```
allow container squashfs = yes      # read only 
allow container extfs = no               # writable container
allow container dir = no                   # sandbox container - chroot  

limit container owners = gmk  # only use containers owned by user

limit container paths = /scratch, /tmp     # only able to use containers in this path

enable overlay = yes        # default is no, bind points within containers 

bind path = /opt/singularity/orca.simg      # bind points extended from host system into the container
```

**Comment by suujia** [comment]
### Documentation 
I will be updating [singularity_manual](https://github.com/bcgsc/orca/blob/master/docs/singularity_manual.md) with proper instructions. 
```
To put files in the container, use ${SINGULARITY_ROOTFS}.
E.g. to put files to container's /usr/local, put it to ${SINGULARITY_ROOTFS}/usr/local
```

**Comment by suujia** [comment]
### Workflow 
```
$ singularity pull --name orca.simg shub://suujia/orca:latest
$ singularity shell --overlay overlay.img orca.simg

Singularity orca.simg:~> java --version
java 9.0.4
Java(TM) SE Runtime Environment (build 9.0.4+11)

Singularity orca.simg:~> brew install nextflow 
```

**Comment by suujia** [comment]
Here's the progress so far for singularityhub. This build has around 300 tools, takes a long time to build but I'm still slowly adding tools to it. I removed the biggest tools to speed up the builds and will add it back later, as well as simplify and clean it up. (Also I will change linuxbrew/.linuxbrew writable actually) 
https://www.singularity-hub.org/containers/3905

Just need to run 
1. singularity pull --name orca.simg shub://suujia/orca:latest 
- this step takes about 3-5 minutes 
2. singularity image.create --size 3000 myoverlay.simg 
- takes ~20 seconds
3.  singularity shell --overlay myoverlay.simg orca.simg
- immediate. launches you in the shell

**Comment by suujia** [comment]
### Final Update 

Brendan fixed some of the permissions and we ended up getting singularity fully on the orca servers fully. It is on permissions 777 because the shub was installing as linuxbrew user and not linuxbrew group. And we can use sudo commands at all for singularity therefore we can't change the identity of user to linuxbrew user to perform the brew install (and we removed the sudo brew install alias). However, the individuals are only modifying/installing things in their own overlay, in their own home directory, which should not be a problem giving the image permissions 777. (You can double check). 

All the tools installed in the big singularity image works. Just need to do a: 
```
singularity pull shub://bcgsc/orca         (add @___COMMIT ID________)
```
into where you'd like the base image to be stored. The thing about singularity hub is that every rebuild overrides the previous, unless you freeze it (similar to tagging it). You can ensure you get that version, even if its same branch, by adding a commit id to the pull. 

Oh the current singularity login script doesn't include the bind (for group folders) that Brendan wanted to add, too.

**Comment by sjackman** [comment]
Excellent. Thanks, Susie!

The current build on Singularity Hub says that it has been running since Tue Aug 28 (one full week). So I'm guessing that something is not quite right there. I'd prefer to build the image on Singularity Hub (or a hosted service), but if that's not possible, we can build it locally and upload the image.
See https://www.singularity-hub.org/collections/1223
